diff --git a/drivers/i2c/i2c_dw.c b/drivers/i2c/i2c_dw.c
index 74cf471783..cc2d11900e 100644
--- a/drivers/i2c/i2c_dw.c
+++ b/drivers/i2c/i2c_dw.c
@@ -40,6 +40,10 @@ LOG_MODULE_REGISTER(i2c_dw);
 
 #include "i2c-priv.h"
 
+
+#define I2C_TRANSFER_TIMEOUT_MSEC		K_MSEC(500)
+
+
 static inline uint32_t get_regs(const struct device *dev)
 {
 	return (uint32_t)DEVICE_MMIO_GET(dev);
@@ -537,8 +541,12 @@ static int i2c_dw_transfer(const struct device *dev,
 		}
 
 		/* Wait for transfer to be done */
-		k_sem_take(&dw->device_sync_sem, K_FOREVER);
-
+		ret = k_sem_take(&dw->device_sync_sem, I2C_TRANSFER_TIMEOUT_MSEC);
+		if (ret != 0) {
+			LOG_ERR("Error on I2C line occurred for message %d", num_msgs - msg_left);
+			ret = -EIO;
+			break;
+		}
 		if (dw->state & I2C_DW_CMD_ERROR) {
 			ret = -EIO;
 			break;
diff --git a/lib/posix/CMakeLists.txt b/lib/posix/CMakeLists.txt
index f9a471d022..5555a04595 100644
--- a/lib/posix/CMakeLists.txt
+++ b/lib/posix/CMakeLists.txt
@@ -12,6 +12,7 @@ if(CONFIG_POSIX_API OR CONFIG_PTHREAD_IPC OR CONFIG_POSIX_CLOCK OR
   # This is a temporary workaround so that Newlib declares the appropriate
   # types for us. POSIX features to be formalized as part of #51211
   zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:-D_POSIX_THREADS>)
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:-DPATH_MAX=256>)
   zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:-D_POSIX_THREADS>)
 endif()
 
